cmd/fossa/cmd/analyze/analyze.go

```diff
@@ -133,6 +133,68 @@ func Do(modules []module.Module) (analyzed []module.Module, err error) {
 }
 
 func uploadAnalysis(normalized []fossa.SourceUnit) error {
+	supportedApiVersions, err := fossa.GetSupportedApiVersions()
+	if err != nil {
+		return err
+	}
+
+	// lists out all compatible versions, where versions listed earlier in the list should be preferred to be used over later appearing versions
+	compatibleVersionsByPreference := []string{"v1", "v0"}
+	for _, compatibleVersion := range compatibleVersionsByPreference {
+		for _, supportedVersion := range supportedApiVersions {
+			if supportedVersion == compatibleVersion {
+				switch supportedVersion {
+				case "v1":
+					return uploadAnalysisV1(normalized)
+				case "v0":
+					fallthrough
+				default:
+					return uploadAnalysisV0(normalized)
+				}
+			}
+		}
+	}
+
+	// default to using the old api when nothing is found
+	log.Warn("Could not find any compatible and supported API versions. Attempting to use the v0 upload path.")
+	return uploadAnalysisV0(normalized)
+}
+
+func uploadAnalysisV1(normalized []fossa.SourceUnit) error {
+	display.InProgress("Uploading analysis...")
+
+	uploadBody := fossa.V1UploadBody{
+		Analysis: normalized,
+	}
+
+	uploadBody.Project.Id = config.Project()
+	uploadBody.Project.Team = config.Team()
+	uploadBody.Project.Url = config.ProjectURL()
+	uploadBody.Project.Title = config.Title()
+	uploadBody.Project.Jira = config.JIRAProjectKey()
+
+	uploadBody.Revision.Link = config.Link()
+	uploadBody.Revision.Id = config.Revision()
+
+	uploadBody.Vcs.Reference = config.Branch()
+
+	if config.GitVcsExists() {
+		uploadBody.Vcs.Type = "git"
+	} else {
+		uploadBody.Vcs.Type = "none"
+	}
+
+	locator, err := fossa.UploadV1(uploadBody)
+	display.ClearProgress()
+	if err != nil {
+		log.Fatalf("Error during upload: %s", err.Error())
+		return err
```

Comment:
This code is unreachable if the line above it logs fatal. I'm not a fan of logging fatal and would prefer to return the error here instead, same with line 216 where this was done before.